### ISSUE TITLE
Always map keys to the same language

### ISSUE THREAD

**Issue by MaximilianGaedig** [Lioness100/guess-that-lang#13]
If you would always map keys to the same language it would remove the extra step of looking down on the options and you could focus on the code.
Great game btw :)

**Comment by MaximilianGaedig** [comment]
This could increase difficulty in guessing the code as there would be much more option, so this could be a higher difficulty, or the 4 options could remain while just showing the different keys they are assigned to

**Comment by sanj0** [comment]
That's exactly what I thought when playing the game and I've been meaning to look into it and give it a go but haven't found the time to implement it yet unfortunately.

**Comment by Lioness100** [comment]
Thanks for opening an issue and showing interest in the game :3 ❤️

This has been suggested by a few other people, but I still don't think I quite understand it. If the languages are always mapped to the same keys, wouldn't there only ever be 4 possible options? (A big downgrade from there currently being 24 languages that you could possibly be quizzed on)

> This could increase difficulty in guessing the code as there would be much more option

Again, wouldn't there be much less options?

> the 4 options could remain while just showing the different keys they are assigned to

Also don't think I quite understand this.

This might be because I just woke up haha and my minds still trying to get back in action, sorry :) could you try to explain in more detail?

**Comment by Lioness100** [comment]
Wouldn't that require the user memorizing every single key? At that point, wouldn't it be easier to just provide a one-line text box where the user can type in either the language name or extension?

**Comment by MaximilianGaedig** [comment]
Yep the player would have to memorise every key assignment, but most people already have a muscle memory of where each key is

**Comment by Lioness100** [comment]
I don't think that would be very beginner friendly

**Comment by MaximilianGaedig** [comment]
You could add it as an option

**Comment by Lioness100** [comment]
I still don't fully understand how it would work. What do you suggest all the keys to be?

**Comment by MaximilianGaedig** [comment]
q - go
w - kotlin
e - javascript
r - typescript
t - rust
.
.
.

**Comment by sanj0** [comment]
@MaximilianGaedig

I've imagined it more like

- g = go
- c = c
- p(y) = python
- j = java
- js = javascript

So that when you see the code snippet and are sure it's c, you wouldn't have to look down into the list to see where c is, but only press the key 'c'. I think the idea with an arbitrary key assignment down the first row on the keyboard wouldn't make the experience better at all.

**Comment by MaximilianGaedig** [comment]
I thought of using the first letter of the language aswell but did not know what to do with conflicts, this is a good idea, typing out the language until there are no other options, also removes the problem of different keyboard layouts

**Comment by sanj0** [comment]
First-letter-collisions would be fairly easy to handle I think as the program would only have to wait ever so shortly after the letter for a second letter to be pressed immediately afterwards. Alternatively, the program could wait for <enter> to avoid frustration for when you lost bc you didn't hit the second key fast enough.

**Comment by MaximilianGaedig** [comment]
Maybe just confirming shorter shortcuts with an enter key press

**Comment by Lioness100** [comment]
I can't really think of a way to implement either of these cleanly. At the very least we'd have to have a key to press to confirm you finished typing your answer, but to have good UX we'd also have to:

- Register all standard text editor keypresses (backspace, arrow keys, delete, etc) which may be more trouble than it's worth
- Provide a good UI in the form a text box so people can see what they're typing

Both of these could be solved using an external crates, but external crates will almost always depend on the cursor being in a predictable position, and under these circumstances, it won't be. We'd have to completely overhaul the terminal integration.

Given the above, I don't think this is a feature I'm going to pursue personally. However, please feel free to submit a PR, I'd love to see what you come up with :)

**Comment by Lioness100** [comment]
Specifically, migrating to [`tui`](https://crates.io/crates/tui) might make this easier, as well as any other additions like this in the future.

**Comment by Lioness100** [comment]
After thinking about it a bit more, I think using `tui` and pursuing this feature is a good idea. Is anybody currently working on this?

**Comment by Lioness100** [comment]
This would also fix resizing issues and allow the language options to be displayed on the side or top
